docs/decisions/0003-initial-api-specification.rst

```diff
@@ -0,0 +1,79 @@
+0003 Initial API Specification
+##############################
+
+Status
+******
+
+**Provisional**
+
+Context
+*******
+
+This doc describes the API specification for the ``subsidy-access-policy`` Django app in this repo.
+
+
+
+Open Questions and TODOs
+========================
+
+
+
+Primary Access Policy Route
+***************************
+**/api/v1/policy/**
+The root URL for all actions allowed on access policy apis.
+
+
+Enterprise Access Policy Transactions
+**************************************
+
+POST access policy redeem transaction
+======================================
+**/api/v1/policy/redeem/**
```

Comment:
This route should probably have the policy uuid as part of the path. `/api/v1/policy/[policy_uuid]/redeem`

Reply:
After thinking through this a little bit, one tricky thing is that there are multiple types of policies encapsulated by different models, e.g. `PerLearnerEnrollmentCapLearnerCreditAccessPolicy` vs. `CappedEnrollmentLearnerCreditAccessPolicy`, and then whatever ones we'll eventually have for licenses. This could make it a little complicated to have a single route like `/api/v1/policy` without knowing exactly which model a view or viewset would need to lookup the PK in.  I'll start a conversation about this.

Reply:
OK, John posted a link/ideas in Slack about trying to fix this up with Proxy models.  To go along with that, here's what I think a more detailed API spec looks like:

The primary rest API endpoint is `redeem` which can accept a GET or a POST.

### Create the redemption

`POST /api/v1/policy/[policy_uuid]/redeem`

This is is scoped to a policy identifier because it saves the computation that likely just happened in preceding GET redeem call (which has to compute against the subsidy and catalog APIs, which is network calls and real work), and also it makes future features of picking amongst multiple policies in a UI more feasible.

Successful POSTs should return an object in the response data containing some identifier indicating that enrollment/entitlement occurred and a URL that the learner-portal interstitial page should redirect to.  We still need to work out the exact implementation details of this.

### Checking redeemability
`GET /api/v1/policy/redeem?enterprise_customer_uuid=[customer_uuid]`

1. Should return a list of policies which could redeem for the given (user, content) pair in the request data.
2. If the (user, content) pair is not redeemable in any policy for the customer, endpoint should return an empty list.
3. Each item in the response results list is the serialized policy, containing at least the course key, access method, catalog id, and priority/preference of the policy.
4. Each item in the list will contain a field indicating the priority/preference of that item.  This makes it clear/explicit from the client perspective how the priority is being chosen.  The results list should be ordered by the values specified in this field.

Why not return a single object or null?  A list is the most generic thing to return, it supports use cases where clients need to know about all redeemable policies out-of-the-box.  It’s also more naturally RESTful (it aligns tightly to the actual data relationships between DB records).